[WIP] Separate Project Histories and Gencode

AnthonyMl opened the pull request:
co-dependent on https://github.com/bdionne/protege-client/pull/6

TODO:
- [ ] remove the literals from the code
- [ ] minimize the path concatenation everywhere or at least its duplication
- [ ] do this for the history files as well
- [ ]  configurable root for the archive so that it can be separated form the current work if needed.

AnthonyMl committed:
Create gencode per project

AnthonyMl committed:
Set evs_history and con_history per project

bdionne commented on src/main/java/org/protege/editor/owl/server/api/ServerLayer.java:
```diff
@@ -49,7 +51,13 @@ public HistoryFile createHistoryFile(String projectId, String projectName) throw
+
+    public void createCodegenFile(String projectId) throws IOException {
+        String rootDir = getConfiguration().getServerRoot() + File.separator + projectId;
+        String filename = rootDir + File.separator + "gencode"; // TODO: remove this constant
```
I think this constant is ok. I can't see a reason to change this name given the file will now be per project

AnthonyMl replied:
I meant to type literal there.
Not changing the name but using a constant instead of a literal.
Just to have it in one place.

bdionne replied:
for sure

bdionne commented:
So one TODO still is to have a second `EVS` history file that only has history from the last time the concept history was generated. When that endpoint is called (it's triggered by the history button in the revision diff panel) the concept history is generated from this monthly `EVS` history file. Then both `EVS` history files and the concept history file are moved to an archive directory. For now I'd just use the date for the directory name and have it as a subdirectory of the project directory. 

Also, let's have a configurable root for the archive so that it can be separated form the current work if needed.

AnthonyMl commented:
Got it.
For the archiving do we want to just move or archive as well (zip/tar/...)?
For the configurable root do you mean a configuration item in ```server-configuration.json```? So instead of ```root/data/projectID``` its ```archive-root/projectID``` where ```archive-root``` is the config item.

bdionne commented:
yes a `serverconfiguration.json` item like `root-archive`, where value might be `/root/archive`. The for each project we'd have `/root/archive/<pid>/<date>/.....` This will allow NCI to specify a remote mount for the archiving.

AnthonyMl committed:
Move evs and concept history files to configurable archive directory

bdionne commented:
This is looking good. We still need:

 * another `evs_history` file, that stays with the project and is cumulative, we keep appending to it. When we generate the concept history, we generate it from the monthly `evs_history`, then archive a copy of all three. Then the monthly `evs_history` is deleted in the project. The generated `con_history` can stay in the project directory, so that it's still accessible. It will get overwritten at the next publication

 * unit tests fail, in the token table there's a null pointer, probably just software atrophy.

bdionne commented on src/main/java/org/protege/editor/owl/server/http/ServerProperties.java:
```diff
@@ -21,6 +21,7 @@
      */
     public static final String EVS_HISTORY_FILE = "evshistory_file";
     public static final String CON_HISTORY_FILE = "conhistory_file";
+    public static final String ARCHIVE_ROOT = "root-archive";
```
total nit: I'd make this name `root_archive` to make life easier on the sysadmin

AnthonyMl replied:
sure

AnthonyMl committed:
Add cur_evs_history file

This file only records the history since the last publication step.
The con_history file is generated from this new file.
Also changed root-archive to root_archive.

AnthonyMl committed:
Use constant instead of literal

AnthonyMl commented:
The TokenTable test is crashing on master as well and seems unrelated.

bdionne commented:
LGTM - runs ok - going to proceed and merge this so we can hopefully tag a final release soon
